From 4e33bf4e26bc6cc8d93b3851fed97c11f513bfe4 Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Fri, 31 May 2019 15:28:00 +0100
Subject: [PATCH] TOOLS-2258 buildimage should drop the zoneroot 'config' dir
 Reviewed by: Chris Burroughs <chris.burroughs@joyent.com> Approved by: Chris
 Burroughs <chris.burroughs@joyent.com>

---
 tools/buildimage/bin/buildimage | 27 ++++++++++++++++++++++++++-
 1 file changed, 26 insertions(+), 1 deletion(-)

diff --git a/tools/buildimage/bin/buildimage b/tools/buildimage/bin/buildimage
index 626ce67..1b7b08f 100755
--- a/tools/buildimage/bin/buildimage
+++ b/tools/buildimage/bin/buildimage
@@ -6,7 +6,7 @@
  */
 
 /*
- * Copyright (c) 2019, Joyent, Inc.
+ * Copyright 2019 Joyent, Inc.
  */
 
 var child_process = require('child_process');
@@ -490,6 +490,30 @@ function pkginCleanup(opts, callback) {
     });
 }
 
+function removeZonerootConfigDir(opts, callback) {
+    // Despite the protections offered by OS-5975 which deletes the
+    // config dir during deployment, we'd like to ensure that images
+    // we create can run safely on PIs that don't have that fix.
+    assert.object(opts, 'opts');
+    assert.string(opts.mountpoint, 'opts.mountpoint');
+    assert.func(callback, 'callback');
+
+    logLine('removing zoneroot config dir ...');
+    const child = child_process.spawn('/usr/bin/rm', [
+        '-rf', opts.mountpoint + '/config'], {
+        stdio: ['ignore', process.stdout, process.stderr]
+    });
+
+    child.on('close', function _onClose(code) {
+        if (code !== 0) {
+            logLine('Child rm zoneroot config dir exited with code ' + code);
+            callback(new Error('failed to remove zoneroot config dir'));
+            return;
+        }
+        callback();
+    });
+}
+
 function doChrootUmount(mountObj, callback) {
     assert.object(mountObj, 'mountObj');
     assert.string(mountObj.dest, 'mountObj.dest');
@@ -702,6 +726,7 @@ function buildImage(opts, callback) {
             loadPkgsrcPkgs,
             pkgAudit,
             pkginCleanup,
+            removeZonerootConfigDir,
             unsetupChroot,
             cleanupZoneAnalog,
             createImage,
-- 
2.21.0

